Chinese remainder theorem

carloshellin opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, **failure to follow the guidelines below will result in the PR being closed automatically**. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access be granted. 🚨

**PLEASE READ THIS**:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject it since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. **Starting April 10, the minimum eval count is 15 samples, we hope this makes it easier to create and contribute evals.**

Also, please note that we're using **Git LFS** for storing the JSON files, so please make sure that you move the JSON file to Git LFS before submitting a PR. Details on how to use Git LFS are available [here](https://git-lfs.com).

## Eval details 📑

### Eval name

Chinese remainder theorem eval

### Eval description

Test the ability of the model to solve systems of congruences with the Chinese remainder theorem

### What makes this a useful eval?

This evaluation tests the model's ability to apply the Chinese remainder theorem. Currently, the Chinese remainder theorem has gained importance because of its application in cryptography, specifically in the RSA system. In this context, the theorem is often used to recover keys.

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [x] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can create an eval on cases where the model fails to reason about the physical world.
- [x] Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.
- [x] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [x] **Include at least 15 high-quality examples.**

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

> Insert what makes your eval high quality that was not mentioned above. (Not required)

## Eval structure 🏗️

Your eval should

- [x] Check that your data is in `evals/registry/data/{name}`
- [x] Check that your YAML is registered at `evals/registry/evals/{name}.yaml`
- [x] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (<https://platform.openai.com/docs/usage-policies>).

- [x] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the commits on the merged pull request.

- [x] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgment

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and the high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [x] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access be granted.

### Submit eval

- [x] I have filled out all required fields of this form
- [x] I have used **Git LFS** for the Eval JSON data
- [ ] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
  {"input":[{"role":"system","content":"Answer the following math question with just the solution. For example: 10x is congruent to 15 modulo 25, response: x is congruent to -1 modulo 5"},{"role":"user","content":"x is congruent to 1 modulo 2\n x is congruent to 4 modulo 7\n x is congruent to 3 modulo 11"}],"ideal":"x is congruent to 25 modulo 154"}
{"input":[{"role":"system","content":"Answer the following math question with just the solution. For example: 10x is congruent to 15 modulo 25, response: x is congruent to -1 modulo 5"},{"role":"user","content":"x is congruent to 0 modulo 11\n x is congruent to 10 modulo 1"}],"ideal":"x is congruent to 43 modulo 77"}
{"input":[{"role":"system","content":"Answer the following math question with just the solution. For example: 10x is congruent to 15 modulo 25, response: x is congruent to -1 modulo 5"},{"role":"user","content":"x is congruent to 5 modulo 6\n x is congruent to 3 modulo 10\n x is congruent to 13 modulo 20"}],"ideal":"x is congruent to 53 modulo 60"}
{"input":[{"role":"system","content":"Answer the following math question with just the solution. For example: 10x is congruent to 15 modulo 25, response: x is congruent to -1 modulo 5"},{"role":"user","content":"x is congruent to 2 modulo 3\n x is congruent to 3 modulo 5\n x is congruent to 2 modulo 7"}],"ideal":"x is congruent to 25 modulo 105"}
{"input":[{"role":"system","content":"Answer the following math question with just the solution. For example: 10x is congruent to 15 modulo 25, response: x is congruent to -1 modulo 5"},{"role":"user","content":"x is congruent to 3 modulo 17\n x is congruent to 4 modulo 11\n x is congruent to 5 modulo 6"}],"ideal":"x is congruent to 785 modulo 1122"}
  ```
</details>

usama-openai reviewed (changes requested):
Thanks for the contribution. I'm getting the following error while evaluating this PR:

```
b'samples = self.get_samples()'
b'File "/content/evals/evals/eval.py", line 150, in get_samples'
b'return get_jsonl(self.samples_jsonl)'
b'File "/content/evals/evals/data.py", line 124, in get_jsonl'
b'return _get_jsonl_file(path)'
b'File "/content/evals/evals/data.py", line 86, in _get_jsonl_file'
b'return [_decode_json(line, path, i + 1) for i, line in enumerate(f)]'
b'File "/content/evals/evals/data.py", line 86, in <listcomp>'
b'return [_decode_json(line, path, i + 1) for i, line in enumerate(f)]'
b'File "/content/evals/evals/data.py", line 80, in _decode_json'
b'raise ValueError(custom_error_message) from None'
b'ValueError: Error parsing JSON on line 15: Invalid \\escape at chinese-remainder-theorem/samples.jsonl:15:272'
```

I would also like to request some changes.

I would recommend asking the model to reason before answering and using the `Includes` evaluation method here because arithmetic operations and other complex tasks are hard for the model to do zero-shot without a chance to reason through the steps. Asking the model to provide reasoning will give the model a fair chance to solve the question. You need to add instructions to provide the chain of thought before answering to give the model a proper chance to reason before answering.

You can update the prompt to something like `Answer the following math question. First, provide your chain of thought, and then provide the final answer. Format your final answer as x is congruent to a modulo b. For example: 10x is congruent to 15 modulo 25, Final answer: x is congruent to -1 modulo 5`.

We would love to review the PR again after the suggested changes.

douglasmonsky commented:
@carloshellin, the first issue raised above is due to a minor mistake in the statement "x is congruent to 3 modulo 8\ x is congruent to 1 modulo 9". There is a bare backslash \ present which is invalid in JSON.

Just as a gentle reminder, in JSON strings, a backslash \ acts as an escape character and needs to be followed by specific characters like \n for newline, \t for tab, \" for a quotation mark, \b for backspace, \f for form feed, \r for carriage return, \u for a unicode character, or another backslash \\ for a literal backslash.

Looking at your other samples, it appears likely you are already aware of this, and it was just a small oversight/typo on this occasion (You meant to put \n instead of just \), so please disregard my comment if that's the case.

Either way, thanks for your contribution to this project.

carloshellin commented:
@usama-openai @douglasmonsky Thank you very much for your suggestions and comments. I have addressed the comments with my last commit.

Best regards,
Carlos

usama-openai reviewed (approved):
This PR looks in good shape now. I'm approving this PR.
